Automattic/sensei
# Add course structure API integration

Architecture:

* A redux store stores the fetched server data and the editor block's state separately. 
* The editor block's `save` function only updates the latter **editor state** in the store. 
* A subscribe on the registry looks for `core/editor`'s `isSavingPost` state, and triggers the store's save function, which POSTs the editor state to the REST API. (When it's not an autosave)
* Then updates the server state, which triggers down to the editor block.

The block's `save` function is called many times during editing (it's basically used as serialize), making it unsuitable to hook up to the **Save Post** intent. 
It's the best place though to extract the structure from the blocks, and is also always called before the post is saved.

The blocks are also saved normally in post_content, which preserves the autosaved draft state on page reload. Unless there is already a non-empty course structure coming from the server, which overwrites it on page load. So a potential for usability improvement here is merging the server state into any existing draft state. 

### Changes proposed in this Pull Request

* Add a store fetching/updating course structure on the REST API
* Add `Edit lesson` button (if they have an ID) or `Unsaved` label to lesson blocks

### Testing instructions

* Open an existing Course with lessons/modules
* Make sure the module and lesson blocks are correct
* Edit things, save post
* Check that the network request contains the updated structure



### Screenshot / Video

![image](https://user-images.githubusercontent.com/176949/92794424-dfee8a80-f3af-11ea-9f96-546aec12ad71.png)
